File: src/test/java/com/giaphi/demo/pages/LoginPage.java
package com.giaphi.demo.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class LoginPage extends BasePage {
    public LoginPage(WebDriver _driver) { // not a method, a constructor (hàm dựng)
        super(_driver);
    }

    // Locators: Id, Css Locator, Xpath
    By byHeader = By.className("login_logo");
    By byUsernameInput = By.id("user-name");
    By byPasswordInput = By.id("password");
    By bySignInButton = By.id("login-button");
    By byErrorMessage = By.cssSelector("h3[data-test='error']");

    // Elements
    public WebElement getHeader() {
        return getElement(byHeader);
    }

    public WebElement getUsernameInput() {
        return getElement(byUsernameInput);
    }

    public WebElement getPasswordInput() {
        return getElement(byPasswordInput);
    }

    public WebElement getLoginButton() {
        return getElement(bySignInButton);
    }

    public WebElement getErrorMessage() {
        return getElement(byErrorMessage);
    }

    // Actions
    public void inputUsername(String username) {
        inputText(getUsernameInput(), username);
    }

    public void inputPassword(String password) {
        inputText(getPasswordInput(), password);
    }

    public InventoryPage clickSignInBtn() {
        this.getLoginButton().click();
        return new InventoryPage(driver);
    }

    public InventoryPage login(String username, String password) {
        inputUsername(username);
        inputPassword(password);
        return clickSignInBtn();
    }
}

File: playwright-saucedemo/pages/login.page.ts
import { Locator, Page } from "@playwright/test";

export class LoginPage {
    private readonly page: Page
    private readonly input_username: Locator
    private readonly input_password: Locator
    private readonly btn_login: Locator
    private readonly message_error: Locator

    constructor(page: Page) {
        this.page = page
        this.input_username = page.locator('#user-name')
        this.input_password = page.locator('#password')
        this.btn_login = page.locator("#login-button")
        this.message_error = page.locator("h3[data-test='error']")
    }
    
    async goto() {
        await this.page.goto("");
    }

    async inputUsername(username: string) {
        await this.input_username.clear()
        await this.input_username.fill(username)
    }

    async inputPassword(password: string) {
        await this.input_password.clear()
        await this.input_password.fill(password)
    }

    async clickLoginButton() {
        await this.btn_login.click()
    }

    async getErrorMessage() {
        return await this.message_error.textContent() ?? ''   
    }
}